### PR TITLE
Validate type hint against requested type

### PR DESCRIPTION
**Pull request by asantas93**
Throw mapping exception when class from type hint is not compatible with the type in the user-specified manifest. https://github.com/json4s/json4s/issues/617

### CODE REVIEW

File: tests/src/test/scala/org/json4s/TypeHintsTest.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  import TypeHintsTest._
+
+  "deserialization with type hints" should {
+
+    implicit val formats: Formats = DefaultFormats + MappedTypeHints(Map(classOf[Foo] -> "foo", classOf[Bar] -> "bar"))
+
+    "fail when the type hint is incompatible with the requested type" in {
+      val dump = Serialization.write(Foo(1))
+      (JsonMethods.parse(dump).extract[Bar]: Any) must throwA[MappingException]
```
Comment (asantas93) [inline tests/src/test/scala/org/json4s/TypeHintsTest.scala:15]: The issue I reported in https://github.com/json4s/json4s/issues/617 only seems to occur when the extracted value is then treated as an instance of `Any` (which is what happened when you pass it to `println`). I don't understand why this changes the behavior of `extract`, but making the change I did to `Extraction` get the behavior I expected originally.

**Comment by mccartney** [comment]
Should we move other tests about `TypeHints` (e.g. from `JsonFormatsSpec`) to this new class?

**Comment by asantas93** [comment]
> @asantas93
> Should we move other tests about `TypeHints` (e.g. from `JsonFormatsSpec`) to this new class?

Well `JsonFormatsSpec` in particular seems to be focused on testing what happens when you merge two `Formats` instances. I could go either way on whether it makes sense to move the `TypeHints`-focused pieces into `TypeHintsTest`.

File: core/src/main/scala/org/json4s/Extraction.scala [hunk 2]
```diff
@@ -667,6 +667,9 @@ object Extraction {
       val deserializer = formats.typeHints.deserialize
       if (!deserializer.isDefinedAt(typeHint, obj)) {
         val concreteClass = formats.typeHints.classFor(typeHint) getOrElse fail("Do not know how to deserialize '" + typeHint + "'")
+        if (!typeInfo.erasure.isAssignableFrom(concreteClass)) {
+          fail(s"Type hint $typeHint which refers to class ${concreteClass.getName} cannot be extracted as an instance of ${typeInfo.erasure.getName}")
```
Comment (asantas93) [inline core/src/main/scala/org/json4s/Extraction.scala:671]: @magnolia-k @mccartney I'm wondering if it would be better to throw a `MappingException` with of the following as the `cause`
1) a `ClassCastException` with a message resembling this
2) a custom exception which holds the 2 incompatible `Class` instances
3) same as `2` but extending `ClassCastException`

Comment (asantas93) [reply core/src/main/scala/org/json4s/Extraction.scala:671]: Alternately, for maximum backward-compatibility,  we could just throw a `ClassCastException`, which is what happens today, assuming the return value of `extract` is stored in a variable declared as the type incompatible with the type hint class.

Comment (mccartney) [reply core/src/main/scala/org/json4s/Extraction.scala:671]: I don't see any preference between the options presented, to be honest.

Comment (asantas93) [reply core/src/main/scala/org/json4s/Extraction.scala:695]: If you have no preference, I think the simple mapping exception seems okay for now. Can we get this PR merged?
